README.md
# scraper

This project scrapes websites.

![Scraper process](doc/scraper.jpg)

The data for this project is stored in elasticsearch. There are four
types (tables in sql-speak). Each of these contain maps
(hashes/dictionaries/documents). They are links (containing urls),
downloaded (web pages that have been scraped), parsed (data extracted,
usually from the leaf pages).

There are two independent processes. The first looks for any link
documents with a state of _new_, and downloads the url and stores
it in the downloaded type. The secon process looks for any downloaded
documents with a state of _new_, and parses it.

Most of the code is general and will work for any site. There is a seed
map fed into the scraping process that contains a few vital pieces of info.
The map should contain the name of the website, which it will use as
the index name. The map may also contain a version (which gets appended to
the index name) and environment (which gets prepended). The map also
has a reference to two functions: one to gather new links from a downloaded
document, and the other to gather parsed information (such as description,
price, etc.) to save into the parsed type.

A document is stored in the error type upon an error from either of these
processes.

### Website configuration

The following is an example configuration for scraping a website.

```clojure
{:name "amazing-site"               ;; will be used as the index name
 :base-url "https://www.amazing-site.com"
 :start-path "/product-categories"  ;; initial page to parse
 :gather-links-fn #'gather-links    ;; function to be called for extracting links
 :gather-info-fn #'gather-info      ;; function to be called for extracting information
 }
```

### Document structure

Each document will contain minimally the following:

```clojure
{:path "the url path of the document (sans the scheme and host)"
 :state "state of this link/downloaded/parsed document. Common states are new, processed, and error."
 :state-updated-on "automatically updated each time the state is updated."
 }
```

Link documents typically contain a :title as well.

Downloaded documents contain all the normal http response fields in addition to the above
(:status, :body, etc.)

Parsed documents supplement the above fields with the extracted data.

## Installation

Download from http://example.com/FIXME.

## Usage

    $ java -jar plain-scrape-0.1.0-standalone.jar [args]

## License

Copyright © 2016 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-11-30
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-11-30
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/plain-scrape/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/plain-scrape/compare/0.1.0...0.1.1


test/scraper/core_test.clj
(ns plain-scrape.core-test
  (:require [clojure.test :refer :all]
            [plain-scrape.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


project.clj
(defproject scraper "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.8.0"]
                 [clj-http "3.4.1"]
                 [clj-time "0.12.2"]
                 [clojurewerkz/elastisch "3.0.0-beta1"]
                 [org.elasticsearch/elasticsearch "2.3.3"]
                 ;; [enlive "1.1.6"]
                 [cheshire "5.6.3"]
                 [reaver "0.1.2"]
                 [semantic-csv "0.1.0"]
                 [org.clojure/data.csv "0.1.3"]
                 ]
  ;; :main ^:skip-aot plain-scrape.core
  :main ^:skip-aot scraper.helpers
  :target-path "target/%s"
  ;; :jvm-opts ["-Xss800m" "-Xms900m"]
  :profiles {:uberjar {:aot :all}})


src/scraper/rexel.clj
(ns scraper.rexel
  (:require [reaver :refer [parse extract-from text attr extract]]
            [scraper.helpers :as h]
            ))

(defn gather-links [parsed]
  (->>
   (concat
    ;; front page links
    (extract-from parsed ".IT33_shop_products span"
                  [:title :path]
                  "a" text
                  "a" (attr "href"))

    ;; category and sub-category page links
    (extract-from parsed "ul dt"
                  [:title :path]
                  "a" text
                  "a" (attr "href"))
    )
   (filter #(and (:title %) (:path %)))
   set
   ))

(defn gather-info [parsed]
  (let [bits
        (concat
         (extract-from parsed ".prodDetailsInfo .prodPrice h2"
                       [:price]
                       "span" text)

         (extract-from parsed ".priceInputCont h4"
                       [:unit-multiple]
                       "abbr" text)

         (extract-from parsed "div.prodDesc > span"
                       [:descript]
                       "h4" text)

         ;; table of detail info
         (into {}
               (map #(vector (:title %) (:val %))
                    (extract-from parsed ".prodInfoTabDet .tabDetailsInfo h1"
                                  [:title :val]
                                  "span:nth-of-type(1)" text
                                  "span:nth-of-type(2)" text
                                  )))
         )]
    (apply merge bits)))
(gather-info parsed)
(-> parsed gather-info keys)
("Material" "Type" :unit-multiple :descript "Opening" "Mfr / Brand" "UPC" "Rexel Part #" "UNSPSC" "Mfr Part #" :price)



(def website {:name "rexel"
              :env :prod
              ;; :version 2
              :base-url "https://www.rexelusa.com"
              :start-path "/"
              :gather-links-fn #'gather-links
              :gather-info-fn #'gather-info
              })






(def prod-fields ["Material" "Type" :unit-multiple :descript "Opening" "Mfr / Brand" "UPC" "Rexel Part #" "UNSPSC" "Mfr Part #" :price])

(def prod-keys (mapv keyword prod-fields))
(defrecord Pricing [upc price descript manuf part-num matnr])

;; (def all-prods
;;   (h/query-string-seq "elliott" "parsed" "_type:parsed"
;;                     {:_source prod-fields
;;                      }))
;; (nth all-prods 0)


(defn pricing-products []
  (->>
   (h/query-string-seq "elliott" "parsed" "_type:parsed"
                       {:_source prod-fields})
   (map :_source)
   (map (apply juxt prod-keys))
   (map #(apply ->Pricing %))
   (map #(let [upc (h/normalize-upc (:upc %))]
           (assoc %
                  :price (-> % :price h/parse-float)
                  :upc upc
                  ;; :matnr (h/upc->matnr upc)
                  )))
   ))

;; (def priced-products
;;   (pricing-products))

;; (def matnrd-products
;;   (doall
;;    (map #(assoc % :matnr (h/upc->matnr (:upc %)))
;;         priced-products)))
;; (take 5 matnrd-products)
;; (last priced-products)
;; (count priced-products)



;; (let [keys prod-keys]
;;   (->>
;;    all-prods
;;    (map :_source)
;;    ;; (map #(assoc % :Unit (-> % :Unit h/parse-float)))
;;    (map (apply juxt keys))
;;    (map #(apply ->Pricing %))
;;    (map #(assoc % :price (-> % :price h/parse-float)))
;;    (map (fn [p]
;;           (let [len (count (:upc p))]
;;             (cond->
;;                 p
;;               (= len 3) (assoc :upc nil)
;;               (= len 14) (assoc :upc (subs (:upc p) 1 13)) ; nix country & check digits
;;               ))))
;;    ;; (map first)
;;    (take 2)
;;    ;; (filter #(= 3 (if (string? %) (count %))))
;;    ;; (map #(if (string? %) (count %)))
;;    ;; (remove nil?)
;;    ;; (take 28790)
;;    ;; set
;;    ))



(h/delete-repository website)
(h/init-repository website)









(def run-atom? (atom true))
(def link-thread (future
                   (h/process-link-thread website run-atom?)))
(def downloaded-thread (future
                         (h/process-downloaded-thread website run-atom?)))
(reset! run-atom? false)



(map
 (fn [x] (-> (h/get-all-next website x) :hits :total))
 ["link" "downloaded" "parsed" "error"])

[@h/links-processed @h/downloaded-processed]









;; (h/check-response-failure website "testing" {:a 3 :b 4} {:response nil} "error while saving")
;; (h/save website "error" {:path "testing"})


;; (gather-info (-> (h/get-id website "downloaded" "/P/Item/PVF/5133710/") :body parse))
;; (h/save-parsed website
;;                (assoc
;;                 (gather-info (-> (h/get-id website "downloaded" "/P/Item/PVF/5133710/") :body parse))
;;                 :path "/P/Item/PVF/5133710/"))

;; (gather-info (h/get-id website "downloaded" "/P/Item/PVF/5133710/"))
;; ;; (:body (h/get-id website "downloaded" "/P/Item/PVF/5133710/"))


;; (h/process-downloaded-thread website run-atom?)
;; (h/process-downloaded-thread website run-atom?)

;; (h/get-id website "link" (:start-path website))
;; (:created-on (h/get-id website "link" (:start-path website)))

;; (h/get-all-next website "link")
;; (h/get-all-next website "downloaded")
;; (-> (h/get-all-next website "downloaded") :hits :hits first :_source)
;; (-> (h/get-all-next website "downloaded") :hits :hits first :_source :state)
;; (h/get-all-next website "parsed")

;; (def x (h/download-url (str (:base-url website) "/P/Category/List/1956-Conduit-Fittings-Bodies-Outlet-Boxes--Accessories")))
;; (gather-links (-> x :body parse))

;; (h/next-unprocessed-link website)
;; (def x (h/next-unprocessed-link website))
;; (h/download-link website x)
;; (def x (h/download-link website x))
;; (keys x)
;; (:body x)
;; (:status x)
;; (:path x)
;; (:start-path website)
;; (h/next-unprocessed-downloaded website)
;; (:body (h/next-unprocessed-downloaded website))

;; ;; (gather-links (h/next-unprocessed-downloaded website))
;; ;; (def z (h/next-unprocessed-downloaded website))
;; ;; (gather-links (-> z :body parse))
;; (gather-info (-> z :body parse))

;; (h/save website "config" {:path "website" :doc website})
;; (h/save-link website {:path (:start-path website)})
;; (map #(h/save-link website %) (-> front-page :body parse gather-links))

;; (h/next-unprocessed-link website)
;; (h/get-path website "link" (:start-path website))
;; (h/get-path website "link" "sss")
;; (h/get-all-next website "link" "new")
;; (h/get-all-next-unprocessed website "link")

;; (h/update-state website "link" (h/get-path website "link" (:start-path website)) "new")

;; (h/get-path website "link" (:start-path website))
;; {:path "/P/Category", :updated-on "2016-12-03T16:40:30Z", :state "new", :created-on "2016-12-03T16:40:30Z", :state-updated-on "2016-12-03T16:40:30Z"}


;; (def front-page (h/download-url (str (:base-url website) (:start-path website))))
;; (def parsed (-> front-page :body parse))
;; (extract-from parsed ".IT33_shop_products span"
;;               [:title :path]
;;               "a" text
;;               "a" (attr "href"))
;; (gather-links parsed)

;; (def category-page (h/download-url "https://www.rexelusa.com/boxes-enclosures/boxes-explosionproof/category/41"))
;; (def parsed (-> category-page :body parse))
;; (extract-from parsed "ul dt"
;;               [:title :path]
;;               "a" text
;;               "a" (attr "href"))
;; (gather-links parsed)

;; (def subcategory-page (h/download-url "https://www.rexelusa.com/boxes-enclosures/boxes-explosionproof/conduit-junction-boxes/category/1167"))
;; (def parsed (-> subcategory-page :body parse))
;; (extract-from parsed "ul dt"
;;               [:title :path]
;;               "a" text
;;               "a" (attr "href"))

;; (def products-page (h/download-url "https://www.rexelusa.com/boxes-enclosures/boxes-explosionproof/conduit-outlet-boxes/type-lb/category/3208"))
;; (def parsed (-> products-page :body parse))
;; (set
;;  (extract-from parsed "#pagination fieldset dl dd"
;;                [:title :path]
;;                "a" text
;;                "a" (attr "href")))



;; gather-info

(def product-page (h/download-url "https://www.rexelusa.com/grss-type/egs/grss100/appleton-grss100-conduit-outlet-box-type-grss-explosionproof-dust-ignitionproof/product/45170"))
(def parsed (-> product-page :body parse))

(let [bits
      (concat
       (extract-from parsed ".prodDetailsInfo .prodPrice h2"
                     [:price]
                     "span" text)

       ;; (clojure.string/split "abc def g" #" ")
       (extract-from parsed ".priceInputCont h4"
                     [:unit-multiple]
                     "abbr" text)

       (extract-from parsed "div.prodDesc > span"
                     [:descript]
                     "h4" text)

       ;; table of detail info
       (extract-from parsed ".prodInfoTabDet .tabDetailsInfo h1"
                     [:title :val]
                     "span:nth-of-type(1)" text
                     "span:nth-of-type(2)" text
                     ))]
  (apply merge bits))


(extract-from parsed ".prodInfoTabDet .tabDetailsInfo h1"
              [:title :val]
              "span:nth-of-type(1)" text
              "span:nth-of-type(2)" text
              )
(into {}
      (map #(vector (:title %) (:val %))
           (extract-from parsed ".prodInfoTabDet .tabDetailsInfo h1"
                         [:title :val]
                         "span:nth-of-type(1)" text
                         "span:nth-of-type(2)" text
                         )))
(keys (into {}
            (map #(vector (:title %) (:val %))
                 (extract-from parsed ".prodInfoTabDet .tabDetailsInfo h1"
                               [:title :val]
                               "span:nth-of-type(1)" text
                               "span:nth-of-type(2)" text
                               ))))
("Mfr / Brand" "Mfr Part #" "Rexel Part #" "UPC" "UNSPSC" "Material" "Opening" "Type")


;; (def front-page (h/download-path website (:start-path website)))
;; (def major-category-page (h/download-url "https://www.elliottelectric.com/P/Category/List/1956-Conduit-Fittings-Bodies-Outlet-Boxes--Accessories"))
;; (def product-page (h/download-url "https://www.elliottelectric.com/P/Item/BRI/106S/"))

;; (->> front-page :body parse (gather-website-links website))
;; (-> front-page :body parse gather-links)
;; (-> major-category-page :body parse gather-links)
;; (-> product-page :body parse product-page-group-names)
;; (-> product-page :body parse gather-stats)




src/scraper/scraper.clj
(ns scraper.scraper
  (:require
            [clojurewerkz.elastisch.rest :as esr]
            ;; [clojurewerkz.elastisch.native :as es]
            ;; [clojurewerkz.elastisch.rest.index :as esi]
            ;; [clojurewerkz.elastisch.rest.index :as esi]
            ;; [clojurewerkz.elastisch.rest.document :as esd]
            ;; [clojurewerkz.elastisch.rest.response :as esrsp]
            ;; [clojurewerkz.elastisch.query :as q]
            ;; [clojure.pprint :as pp]

            ;; [clj-time.core :as t :refer [now]]
            ;; [clj-time.coerce :as c]
            ;; ;; [clj-time.local :as l]

            [scraper.helpers :as h]
            ))

;; the repository will have the following tables:
;;   - link, for the links that are to be downloaded.
;;     this list grows as pages have links to other pages
;;   - downloaded, for links that have been downloaded
;;   - processed, for downloaded pages that have been processed
;;
;; each document in the above tables has a :state variable.
;; the state variable can be:
;;   - :new (has not been processed)
;;   - :processing (is currently being processed)
;;   - :processed (has completed processing)
;;   - :error (in which case, :error-reason will have more info)

(def example-link
  {:path "/a/b"
   :state :new  ; :processing :processed :error-downloading :error-parsing
   :state-updated-on "now"
   :requested-on "..."
   :downloaded-on "..."
   :parsed-on "..."
   :body "......."
   :sublinks ["/a/b/c" "/c"]
   :vitals {:upc ".." :price 2.43 :uom "EA"}
   }
  )

(def example-page
  {:path "/a/b"
   ;; :
   })




;; (def website "elliott")
;; (def repo-url "http://127.0.0.1:9200")
;; (def repo
;;   (esr/connect
;;    repo-url
;;    {:connection-manager
;;     (clj-http.conn-mgr/make-reusable-conn-manager {:timeout 10})}))

;; (h/save  repo website "downloaded" {:path "/ab/cd" :a 9 :q 99})
;; (h/save  repo website "downloaded" {:path "/ab/cd" :a 17})
;; (esd/get repo website "downloaded" "/ab/cd")
;; (h/next-unprocessed-link repo website)
;; (h/next-unprocessed-downloaded repo website)

;; {:_index "elliot", :_type "downloaded", :_id "/ab/cd", :_version 3, :found true, :_source {:path "/ab/cd", :a 17, :state "new", :state-updated-on "2016-12-01T21:14:00Z"}}



(defn update-as-processing
  [page]
  )

(defn process-next-downloaded
  [website]
  (let [link (h/next-unprocessed-link repo website)]
    ))

(defn process-all-downloaded
  [website]
  (loop []
      (when (process-next-downloaded website)
        (recur))))

(defn gather-links
  [website link page]
  )

(defn download-next [website]
  (let [link (h/next-unprocessed-link repo website)
        page (if link (h/download-url (h/build-url website link)))]
    (try
      (when page
        (h/save-downloaded repo website link page)
        (h/update-state repo website link :processed)
        (try
          (gather-links website link page)
          (catch Exception e
            (h/update-state repo website link :error
                          {:error-state :gathering-links :error-msg (str e)}))))
      (catch Exception e
        (h/update-state repo website link :error
                      {:error-state :gathering-links :error-msg (str e)}))
      )
    link)
    )

(defn crawl
  "crawl the site, downloading pages as it goes"
  [website start-page process-fn options]
  (h/init-repository repo website)
  (let [opts (merge
              {:delay-ms 2000
               :process-downloaded true
               }
              options)]
    (loop []
        (when (download-next website)
          (Thread/sleep (:delay-ms opts))
          (when (:process-downloaded options)
            (process-next-downloaded website)
            (process-next-downloaded website))
          (recur)))))


src/scraper/helpers.clj
(ns scraper.helpers
  (:require [clojurewerkz.elastisch.rest :as esr]
            [clojurewerkz.elastisch.rest.index :as esi]
            [clojurewerkz.elastisch.rest.document :as esd]
            [clojurewerkz.elastisch.rest.response :as esrsp]
            [clojurewerkz.elastisch.query :as q]
            [clojure.pprint :as pp]

            [reaver :refer [parse extract-from text attr extract]]
            [clj-http.client :as client]
            ))

(def repo-url (if-let [url (System/getenv "ELASISEARCH_URL")]
                url
                ;; "http://127.0.0.1:9200"
                "http://192.168.0.220:9201"
                ;; "http://murphydye.com:9200"
                ))
(def repo
  (esr/connect
   repo-url
   {:connection-manager
    (clj-http.conn-mgr/make-reusable-conn-manager {:timeout 10})}))

(defn to-s [obj]
  (condp contains? (type obj)
    #{clojure.lang.Keyword clojure.lang.Symbol java.lang.String} (name obj)
    #{Long Integer} (str obj)
    (str obj)))
;; (to-s "a")
;; (to-s :a)
;; (to-s 'a)
;; (to-s 2)

(defn website-index-name [website]
  (if (map? website)
    (clojure.string/join "_" (map to-s (remove nil? [(:name website) (:env website) (:version website)])))
    (to-s website)))
;; (website-index-name {:name "boo" :version 2 :env :prod})

(defn download-url
  "return response from url's content"
  ([url] (download-url url {:socket-timeout 10000 :conn-timeout 1000}))
  ([url options]
   (let [response (client/get url options)]
     response)))
;; (download-url "http://google.com/")

(defn download-link
  ([website link] (download-link website link {}))
  ([website link options]
   (prn website)
   (download-url (str (:base-url website) (:path link)) options)))





;; -------------------------------  repository document functions

(defn timenow []
  (java.util.Date.))



(defn get-first-doc [response]
  (-> response :hits :hits first :_source))

(defn get-id [website type id]
  (let [index (website-index-name website)
        response (esd/get repo index type id)]
    (when response
      (:_source response))))

(defn exists? [website type doc]
  (if (get-id website type (:path doc))
    true
    false))

(defn get-all-next
  ([website type]
   (esd/search repo (website-index-name website) type
               {:query {:match_all {}}
                :size 1}))
  ([website type state]
   (esd/search repo (website-index-name website) type
               {:query {:match {:state state}}
                :size 1})))

(defn query-all-seq
  ([website type] (query-all-seq website type {}))
  ([website type options]
   (esd/scroll-seq repo
                   (esd/search repo (website-index-name website) (name type)
                               (merge
                                {:query {:match_all {}}
                                 :scroll "1m"
                                 :size 1000}
                                options)))))
;; (take 2 (query-all-seq "ecommerce" "product"))

(defn query-string-seq
  ([website type qstring] (query-string-seq website type qstring {}))
  ([website type qstring options]
   (if (empty? qstring)
     (query-all-seq website type options)
     (let [response
           (esd/scroll-seq repo
                           (esd/search repo (website-index-name website) (name type)
                                       (merge
                                        {:query {:query_string {:query qstring}}
                                         :scroll "1m"
                                         :size 1000}
                                        options)))]
       response))))

;; (take 2 (esd/scroll-seq repo
;;          (esd/search repo "ecommerce" "product"
;;                      {:query {:query_string {:query "wire"}}
;;                       :scroll "1m"
;;                       :size 1000})))
;; (take 2 (query-string-seq "ecommerce" "product" "wire"))
;; (take 2 (query-string-seq "ecommerce" "product" ""))
;; (take 2
;;       (let [query-string "wire"
;;             website "ecommerce"]
;;         (esd/scroll-seq repo
;;                         (esd/search repo (website-index-name website) type
;;                                     {:query {:query_string {:query query-string}}
;;                                      :scroll "1m"
;;                                      :size 1000}
;;                                     ))))

(defn parse-long [s]
  (if (string? s)
    (if-let [token (re-find #"\d+" s)]
      (if (string? token)
        (Long. token)))))
(defn parse-float [s]
  (if (string? s)
    (if-let [token (re-find #"\d+\.?\d+" s)]
      (if (string? token)
        (Double. token)))))
;; (parse-float "$2.32 ea")
;; (parse-float "$2.32 ea 44.2")

(defn query-string
  ([website type qstring] (query-string website type qstring {}))
  ([website type qstring options]
   (let [response
         (esd/search repo (website-index-name website) type
                     (merge
                      {:query {:query_string {:query qstring}}
                       :size 2}
                      options))
         ]
     response)))
;; (query-string "ecommerce" "product" "upc:\"032664576553\"" {})

(defn query-string-and-hits
  ([website type qstring] (query-string-and-hits website type qstring {}))
  ([website type qstring options]
   (let [response (query-string website type qstring options)]
     [(-> response :hits :total)
      (->>
       response
       :hits :hits
       (map :_source)
       (map #(dissoc % :error-state))
       )])))

(defn query-string-docs
  ([website type qstring] (query-string website type qstring {}))
  ([website type qstring options]
   (second (query-string-and-hits website type qstring options))))

(defn normalize-upc [upc]
  (when upc
    (let [len (count upc)]
      (cond
        (= len 3) nil
        ;; (= len 13) (subs upc 0 12) ; check digits
        ;; (= len 14) (subs upc 1 13) ; nix country & check digits
        (= len 13) (subs upc 1 13) ; check digits
        (= len 14) (subs upc 2 14) ; nix country & check digits
        :else upc
        ))))
;; (normalize-upc "032664576553")
;; (normalize-upc "0326645765534")
;; (normalize-upc "00326645765534")

(defn upc->matnr [upc]
  (when upc
    (->
     (query-string "ecommerce" "product" (str "upc:" upc))
     :hits :hits first :_source :matnr parse-long
     )))
;; (upc->matnr "032664576553")    ;; has two matnrs!
;; (upc->matnr "786685682846")
;; (upc->matnr "008870053894")



(defn get-all-next-unprocessed
  [website type]
  (get-all-next website type "new"))

(defn get-next-unprocessed
  [website type]
  (let [one (get-all-next-unprocessed website type)]
    (if (<= 1 (-> one :hits :total))
      (-> one :hits :hits first :_source)
      nil
      )))

(defn next-unprocessed-link
  [website]
  (get-next-unprocessed website "link"))

(defn next-unprocessed-downloaded
  [website]
  (get-next-unprocessed website "downloaded"))



(declare save-error)

(defn check-response-failure
  [website typee doc response msg]
  (let [failed (-> response :_shards :failed)
        right-type? (contains? #{java.lang.Long java.lang.Integer} (type failed))]
    (when (or (not right-type?) (< 0 failed))
      (println "error, check-response-failure" typee doc response msg)
      (save-error website {:msg msg :type typee :doc doc :response response})
      )))

(defn save
  [website type doc]
  {:pre [(contains? doc :path)]}
  (let [index (website-index-name website)
        now (timenow)
        response (esd/put repo index type (:path doc)
                 (assoc
                  (cond-> doc
                    (not (contains? doc :created-on)) (assoc :state :new :created-on now)
                    (not (contains? doc :state)) (assoc :state :new :state-updated-on now)
                    )
                  :updated-on now))]
    (def x2 response)
    (check-response-failure website type doc response "error while saving")
    response))


(defn create
  [website type doc]
  (let [index (website-index-name website)
        now (timenow)
        response (esd/create repo index type
                    (assoc
                     (cond-> doc
                       (not (contains? doc :created-on)) (assoc :state :new :created-on now)
                       )
                     :updated-on now))]
    (check-response-failure website type doc response "error while creating")
    response))




(defn save-error
  [website doc]
  (create website "error" doc))

(defn save-new
  [website type doc]
  (when (nil? (exists? website type doc))
    (save website type doc)))

(defn save-link
  [website doc]
  (save website "link" doc))

(defn create-link
  [website doc]
  (when-not (exists? website "link" doc)
    (save-link website doc)))

(defn save-downloaded
  [website doc]
  (save website "downloaded" doc))

(defn create-downloaded
  [website doc]
  (when-not (exists? website "downloaded" doc)
    (save-downloaded website doc)))

(defn save-parsed
  [website doc]
  (save website "parsed" doc))

(defn update-state
  ([website type doc new-state] (update-state website type doc new-state {}))
  ([website type doc new-state options]
   {:pre (contains? doc :path)}
   (save website type
         (merge
          options
          (assoc doc :state new-state :state-updated-on (timenow))))
   )
  )


;; ---------------------------- processing

(defn process-link
  [website link]
  1 / 0
  (try
    (let [response (download-link website link)]
      (if (= (:status response) 200)
        (do
          (create-downloaded website (merge
                                    response
                                    {:path (:path link)}
                                    ))
          (update-state website "link" link :processed))
        (do
          (update-state website "link" link :error {:error-msg "download error" :error-response response})
          (save-error website {:fn :process-link :doc link :response response}))
        )
      )
    (catch Exception e
      (println "error in process-link" link (str e))
      (update-state website "link" link :error {:error-state (str e)})
      (save-error website {:fn :process-link :doc link :msg (str e)})
      :error
      )))

(defn process-downloaded
  [website downloaded]
  (try
    (let [parsed (-> downloaded :body parse)
          links ((:gather-links-fn website) parsed)
          info ((:gather-info-fn website) parsed)
          ]
      (def lll links)
      (def mmm (map #(create-link website %) links))
      (doall mmm)
      (when info (save-parsed website (assoc info :path (:path downloaded))))
      (update-state website "downloaded" downloaded :processed)
      )
    (catch Exception e
      (println "error in process-downloaded" downloaded (str e))
      (update-state website "downloaded" downloaded :error {:error-state (str e)})
      (save-error website {:fn :process-downloaded :doc downloaded :msg (str e)})
      )))

;; lll
;; mmm

(defn process-next-link
  [website]
  (when-let [l (next-unprocessed-link website)]
    (process-link website l)))


(defn process-next-downloaded
  [website]
  (when-let [d (next-unprocessed-downloaded website)]
    (process-downloaded website d)))

(defonce links-processed (atom 0))
(defonce downloaded-processed (atom 0))

;; [@links-processed @downloaded-processed]

(defn process-link-thread
  [website run-atom?]
  (while @run-atom?
    (swap! links-processed inc)
    (process-next-link website)
    (Thread/sleep 4000)
    ))
(defn process-link-thread
  [website run-atom?]
  (while @run-atom?
    (swap! links-processed inc)
    (process-next-link website)
    (Thread/sleep 3800)
    ))

(defn process-downloaded-thread
  [website run-atom?]
  (loop []
    (swap! downloaded-processed inc)
    (process-next-downloaded website)
    (Thread/sleep 2000)
    (if @run-atom? (recur))))


;; ----------------------------- repository index functions

(defn init-repository
  "initialize scraper index. throw error if unable to create it (and it did not already exist)"
  [website]
  {:pre [(contains? website :name)
         (contains? website :start-path)]}
  (try
    (esi/create repo (website-index-name website))
    (catch Exception e
      (println "index already exists" e)
      (when-not (->> e ex-data :body (re-find #"already exists"))
        (throw e))))
  (if (exists? website "link" {:path "start-path"})
    :start-path-already-existed
    (do
      (save-link website {:title "start-path" :path (:start-path website)})
      ;; :created
      )))

(defn delete-repository
  "delete scraper index. throw error if unable to delete it (and it didn't exist)"
  [website]
  {:pre [(contains? website :name)]}
  (try
    (let [response (esi/delete repo (website-index-name website))]
      (cond
        (= (:acknowledged response) true) :deleted
        (= (-> response :error :type) "index_not_found_exception") :did-not-exist
        :else (throw (str "error deleting repository " website))
        ))
    (catch Exception e
      (throw e)
      )))





;; ------------------------------- errors

(defn delete-errors
  "delete from index-name 'error' with fn:link-process"
  [website query-string]
  (let [index (website-index-name website)
        error-ids
        (map :_id (-> (query-string index "error" query-string {:size 9999}) :hits :hits))]
    (map #(esd/delete repo index "error" %) error-ids)))

(defn retry-link-errors
  "delete-error-links first"
  [website]
  (delete-errors website "fn:link-process")
  (let [index (website-index-name website)
        link-ers
         (query-string-docs index
                            "link" "+state:error" {:size 2000})]
    (doall
     (map #(save index "link" (assoc % :state "new")) link-ers))
     ))
;; (count link-ers)
;; (def link-errs
;;   (set
;;    (map :path (query-string-docs "elliott" "link" "state:error" {:size 2000}))))
;; (count link-errs)

;; (def error-ids
;;   (set
;;    (map :_id (-> (query-string "elliott" "error" "fn:link-process" {:size 2000}) :hits :hits))))
;; (def error-links
;;   (set
;;    (map (comp :path :doc) (query-string-docs "elliott" "error" "fn:link-process" {:size 2000}))))
;; (count error-links)
;; (clojure.set/difference link-errs error-links)
;; (clojure.set/difference error-links link-errs)
;; (= error-links link-errs)

;; (doall
;;  (map #(esd/delete repo "elliott" "error" %) error-ids))

;; (doall
;;  (map #(esd/update
;;         )))

;; (esd/put repo "elliott" "error" "testing" {:a 3 :b 4})
;; (esd/put repo "elliott" "error" "testing" {:c 5})
;; (esd/get repo "elliott" "error" "testing")

;; (count (query-string-docs "elliott" "error" "fn:link-process" {:size 30}))
;; (query-string-and-hits "elliott" "error" "fn:link-process")



;; ------------


;; (defn slink [query-string]
;;   (let [response
;;         (esd/search repo "elliott" "link"
;;                     {:query {:query_string {:query query-string}}
;;                      ;; :size 1
;;                      })]
;;     [(-> response :hits :total)
;;      (->>
;;       response
;;       :hits :hits
;;       (map :_source)
;;       (map #(dissoc % :error-state))
;;       ;; (map :path)
;;       )]))

;; (slink "state:error")

;; (def link-errs
;;   (->> (slink "state:error")
;;        second
;;        (map :path)
;;        ))
;; (count link-errs)


;; (slink "state:error AND path:category")
;; (slink "state:error AND path:'category list'")
;; (slink "state:error AND path:'category list 1956 conduit 10 outlet boxes'")
;; (-> (slink "state:error AND path:'category list 1956 conduit 10 outlet boxes'") second first :path)
;; (def pp (-> (slink "state:error AND path:'category list 1956 conduit 10 outlet boxes'") second first :path))
;; (get-id {:name "elliott"} "link" p)
;; (esd/get repo "elliott" "link" p)
;; "/P/Category/List/1956-Conduit-Fittings-Bodies-Outlet-Boxes--Accessories?CurrentPage=10"

;; (gid "elliott" "link" p)
;; (:_source (gid "elliott" "link" p))
;; (defn gid [website type id]
;;   (esd/get repo website type id))


;; (->>
;;  (esd/search repo "elliott" "link"
;;              {:query {:query_string {:query "state:error AND path:category"}}
;;               ;; :size 1
;;               })
;;  :hits :hits
;;  (map :_source)
;;  (map #(dissoc % :error-state))
;;  ;; (map :path)
;;  )


;; (->
;;  (esd/search repo "elliott" "link"
;;              {:query {:match {:state "error"
;;                               ;; :path "category"
;;                               }}
;;               :size 1})
;;  :hits :hits first :_source
;;  (dissoc :error-state)
;;  )

;; (->
;;  (esd/search repo "elliott" "error"
;;              {:query {:match {:fn :process-link}}
;;               :size 1})
;;  :hits :hits first :_source :doc :path)

;; (esd/search repo "elliott" "link"
;;             {:query {:match {:fn :process-link}}
;;              :size 1})

;; (def p "/P/Category/List/1956-Conduit-Fittings-Bodies-Outlet-Boxes--Accessories?CurrentPage=10")
;; (get-id {:name "elliot"}
;;         "link"
;;         p)


